### PR TITLE
Version Management ProjektINF

### PR DESCRIPTION
**Pull request by Ruben-Verma**
This pull request adds  version management to winery. It enables an automated update of a node template. It enables the automatic transfer of the old Key Value Property from the outdated nodeTemplate to the new. The user can map the value of removed and new properties. 

Exclamation mark appears when any new versions are available.
Bildschirmfoto 2019-05-08 um 14 48 12

After clicking the exclamation mark, a modal with newer versions appears which can be chosen.
Bildschirmfoto 2019-05-08 um 14 50 45

Removed,new and resolved properties are shown. User can map new and removed properties.
Bildschirmfoto 2019-05-08 um 14 51 53

Node template is now updated. 
Bildschirmfoto 2019-05-08 um 14 52 42

Update in the UI is supported  for node templates only. 

- [x] Ensure that you followed http://eclipse.github.io/winery/dev/ToolChain#github---prepare-pull-request.
- [x] Branch name checked. That means, the branch name starts with `thesis/`, `fix/`, ...
- [x] Ensure that the commit message is [a good commit message](https://github.com/joelparkerhenderson/git_commit_message)
- [x] Ensure to use auto format in **all** files
- [x] Ensure that you appear in `NOTICE` at Copyright Holders
- [ ] Tests created for changes
- [x] Documentation updated (if needed)
- [x] Screenshots added (for UI changes)

### CODE REVIEW

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/datatypes/PropertiesComparison.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+/*******************************************************************************
+ * Copyright (c) 2012-2013 Contributors to the Eclipse Foundation
```
Comment (saatkamp) [inline org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/datatypes/PropertiesComparison.java]: why do you need this class? remove it.

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/datatypes/UpdateInfo.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
+package org.eclipse.winery.repository.rest.datatypes;
```
Comment (saatkamp) [inline org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/datatypes/UpdateInfo.java]: Where is the copyright?

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/servicetemplates/ServiceTemplateResource.java [hunk 3]
```diff
@@ -355,4 +364,57 @@ protected TExtensibleElements createNewElement() {
...
+                return nullList;
+            
+        }
+        
+        List<TNodeTemplate> nodeList = element.getTopologyTemplate().getNodeTemplates();
+        List<TRelationshipTemplate> relationList = element.getTopologyTemplate().getRelationshipTemplates();
+
+        Map<QName, List<WineryVersion>> versionElements = new HashMap<>();
+
+       /* for (TRelationshipTemplate relation : relationList) {
```
Comment (saatkamp) [inline org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/servicetemplates/ServiceTemplateResource.java]: if you don't need this code, remove it.

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/servicetemplates/topologytemplates/TopologyTemplateResource.java [hunk 4]
```diff
@@ -15,8 +15,8 @@
 
 import java.io.IOException;
 import java.net.URI;
-import java.util.ArrayList;
-import java.util.List;
+import java.util.*;
```
Comment (saatkamp) [inline org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/servicetemplates/topologytemplates/TopologyTemplateResource.java]: You did not use the correct intelliJ settings. * imports are not allowed.

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/servicetemplates/topologytemplates/TopologyTemplateResource.java [hunk 5]
```diff
@@ -292,6 +298,153 @@ public Response resolve(@Context UriInfo uriInfo) {
         }
     }
 
+    @POST
+    @Path("kvcomparison")
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    public PropertyDiffList computeDi(UpdateInfo updateInfo) {
+        // TNodeTemplate is sub class of TEntity, which has Properties. Properties has the method to get all KV-properties
```
Comment (saatkamp) [inline org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/servicetemplates/topologytemplates/TopologyTemplateResource.java]: remove unnecessary comments

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/servicetemplates/topologytemplates/TopologyTemplateResource.java [hunk 6]
```diff
@@ -292,6 +298,153 @@ public Response resolve(@Context UriInfo uriInfo) {
...
+
+            IRepository repository = RepositoryFactory.getRepository();
+            TEntityType tNodeTypeOfNewType = repository.getElement(new NodeTypeId(qNameType));
+            
+            newKvs = tNodeTypeOfNewType.getWinerysPropertiesDefinition()
+                .getPropertyDefinitionKVList()
+                .getPropertyDefinitionKVs();
+            
+            /*
+            resolvedProperties wird erstellt indem man die Keys(Properties) der newKvs von der PropertieDefinitionKV extrahiert
```
Comment (saatkamp) [inline org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/servicetemplates/topologytemplates/TopologyTemplateResource.java]: do you need this code?

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/servicetemplates/topologytemplates/TopologyTemplateResource.java [hunk 7]
```diff
@@ -292,6 +298,153 @@ public Response resolve(@Context UriInfo uriInfo) {
...
+            resolvedProperties = newKvs.stream()
+                .map(PropertyDefinitionKV::getKey)
+                .filter(keys ->  oldKvs.keySet()
+                                .stream()
+                                .map(String::toLowerCase)
+                                .collect(Collectors.toList()).contains(keys.toLowerCase()))
+                .collect(Collectors.toList());
+            
+            /*
+            removedProperties wird erstellt indem man die alten Keys in oldKvs mit den Keys in newKvs vergleicht ( die Liste der neuen Keys wird
```
Comment (saatkamp) [inline org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/servicetemplates/topologytemplates/TopologyTemplateResource.java]: do you need this code?

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/servicetemplates/topologytemplates/TopologyTemplateResource.java [hunk 8]
```diff
@@ -292,6 +298,153 @@ public Response resolve(@Context UriInfo uriInfo) {
...
+             */
+            
+            removedProperties = oldKvs.keySet().stream()
+                .filter(keys -> !newKvs.stream()
+                                .map(newProp -> newProp.getKey().toLowerCase())
+                                .collect(Collectors.toList()).contains(keys.toLowerCase()))
+                .collect(Collectors.toList());
+            
+            /*
+            newProperties wird erstellt indem man die neuen Keys in newKvs mit den alten Keys in oldKvs vergleicht. Ist der neue key nicht in oldKvs vorhanden
```
Comment (saatkamp) [inline org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/servicetemplates/topologytemplates/TopologyTemplateResource.java]: do you need this code?

File: org.eclipse.winery.repository.rest/src/test/java/org/eclipse/winery/repository/rest/resources/AbstractResourceTest.java [hunk 9]
```diff
@@ -74,7 +74,7 @@ public static String readFromClasspath(String fileName) {
     protected RequestSpecification start() {
         return given()
             .log()
-            .ifValidationFails();
+            .all();
```
Comment (saatkamp) [inline org.eclipse.winery.repository.rest/src/test/java/org/eclipse/winery/repository/rest/resources/AbstractResourceTest.java]: undo this for the PR please in the whole file.

File: package-lock.json [hunk 10]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (lharzenetter) [inline package-lock.json]: do not checkin this file

**Comment by Ruben-Verma** [comment]
Follow up https://github.com/eclipse/winery/pull/410

**Pull request closed without merging**